src/sinks/util/statistic.rs

```diff
@@ -21,60 +22,86 @@ pub struct DistributionStatistic {
 
 impl DistributionStatistic {
     pub fn from_samples(source: &[Sample], quantiles: &[f64]) -> Option<Self> {
-        let mut samples = Vec::new();
-        for sample in source {
-            for _ in 0..sample.rate {
-                samples.push(sample.value);
-            }
-        }
+        let mut bins = source
+            .iter()
+            .filter(|sample| sample.rate > 0)
+            .cloned()
+            .collect::<Vec<_>>();
 
-        if samples.is_empty() {
+        if bins.is_empty() {
             return None;
         }
 
-        if samples.len() == 1 {
-            let val = samples[0];
+        if bins.len() == 1 {
+            let val = bins[0].value;
+            let count = bins[0].rate;
             return Some(Self {
                 min: val,
                 max: val,
                 median: val,
                 avg: val,
-                sum: val,
-                count: 1,
+                sum: val * count as f64,
+                count: count as u64,
                 quantiles: quantiles.iter().map(|&p| (p, val)).collect(),
             });
         }
 
-        samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
+        bins.sort_unstable_by(|a, b| a.value.partial_cmp(&b.value).unwrap_or(Ordering::Equal));
+
+        let min = bins.first().unwrap().value;
+        let max = bins.last().unwrap().value;
+        let sum = bins
+            .iter()
+            .map(|sample| sample.value * sample.rate as f64)
+            .sum::<f64>();
+        let count = bins.iter().map(|sample| sample.rate).sum::<u32>();
```

Comment:
This can be calculated from `bins.last().unwrap().rate` if you move it below the loop below.